server/src/main/java/com/exacaster/lighter/application/sessions/SessionHandler.java

```diff
@@ -67,7 +69,9 @@ public void keepPermanentSession() {
             return;
         }
         var session = sessionService.fetchOne(sessionId);
-        if (session.isEmpty() || session.filter(it -> it.getState().isComplete()).isPresent()) {
+        var running = not(ApplicationState::isComplete);
+        if (session.map(Application::getState).filter(running).isEmpty() ||
+                backend.getInfo(session.get()).map(ApplicationInfo::getState).filter(running).isEmpty()) {
```

Comment:
`session.get()` can return null. Maybe we should check `session.isPresent()` too?

Reply:
or even:
`session.map(backend::getInfo).map(ApplicationInfo::getState)....`

Reply:
> `session.get()` can return null. Maybe we should check `session.isPresent()` too?

session.get() can't be null here - checked before OR.

Reply:
before OR you checked if `running` is empty, not original session.

Reply:
> before OR you checked if `running` is empty, not original session.

both. if session is empty before filter `running` - it remains empty anyway :)

Reply:
oh, right 🤦